From d922f5224e539d2a84c8ab9dd828bd344a5601a4 Mon Sep 17 00:00:00 2001
From: Chris <meischris39@hotmail.com>
Date: Thu, 15 Oct 2026 18:02:56 +0100
Subject: [PATCH] Save the Neo Geo's pd4990a clock in savestates

None of the calendar chip's state was registered, so loading a state
left the clock, its serial shift registers and its test pulse where
they were. Run-ahead loads a state every frame, and a Neo Geo then
stalls on boot or stops at CALENDAR ERROR.
---
 src/machine/neogeo_machine.c |  2 ++
 src/machine/pd4990a.c        | 35 +++++++++++++++++++++++++++++------
 2 files changed, 31 insertions(+), 6 deletions(-)

diff --git a/src/machine/neogeo_machine.c b/src/machine/neogeo_machine.c
index f5067fa..28b9b6e 100644
--- a/src/machine/neogeo_machine.c
+++ b/src/machine/neogeo_machine.c
@@ -883,6 +883,8 @@ static void neogeo_register_sub_savestate(void)
 {
 	data8_t* gamevector = memory_region(REGION_CPU1);
 
+	pd4990a_init();
+
 	state_save_register_int   ("neogeo", 0, "sram_locked",             &sram_locked);
 	state_save_register_UINT16("neogeo", 0, "neogeo_ram16",            neogeo_ram16,             0x10000/2);
 	state_save_register_UINT8 ("neogeo", 0, "neogeo_memcard",          neogeo_memcard,           0x800);
diff --git a/src/machine/pd4990a.c b/src/machine/pd4990a.c
index 86bc38c..692d1e5 100644
--- a/src/machine/pd4990a.c
+++ b/src/machine/pd4990a.c
@@ -32,6 +32,7 @@
 
 #include "driver.h"
 #include "machine/pd4990a.h"
+#include "state.h"
 
 
 /* Set the data in the chip to Monday 09/09/73 00:00:00 	*/
@@ -47,7 +48,7 @@ struct pd4990a_s pd4990a =
 	1		/* weekday BCD */
 };
 
-static unsigned int shiftlo,shifthi;
+static UINT32 shiftlo,shifthi;
 
 static int retraces = 0;	/* Assumes 60 retraces a second */
 static int testwaits= 0;
@@ -56,13 +57,38 @@ static int testbit = 0;	/* Pulses a bit in order to simulate */
 					/* test output */
 static int outputbit=0;
 static int bitno=0;
-static char reading=0;
-static char writting=0;
+static int reading=0;
+static int writting=0;
+static int clock_line=0;
+static int command_line=0;	/*??*/
 
 #define DATA_BIT	0x1
 #define CLOCK_BIT	0x2
 #define END_BIT		0x4
 
+void pd4990a_init(void)
+{
+	state_save_register_int   ("pd4990a", 0, "seconds",      &pd4990a.seconds);
+	state_save_register_int   ("pd4990a", 0, "minutes",      &pd4990a.minutes);
+	state_save_register_int   ("pd4990a", 0, "hours",        &pd4990a.hours);
+	state_save_register_int   ("pd4990a", 0, "days",         &pd4990a.days);
+	state_save_register_int   ("pd4990a", 0, "month",        &pd4990a.month);
+	state_save_register_int   ("pd4990a", 0, "year",         &pd4990a.year);
+	state_save_register_int   ("pd4990a", 0, "weekday",      &pd4990a.weekday);
+	state_save_register_UINT32("pd4990a", 0, "shiftlo",      &shiftlo, 1);
+	state_save_register_UINT32("pd4990a", 0, "shifthi",      &shifthi, 1);
+	state_save_register_int   ("pd4990a", 0, "retraces",     &retraces);
+	state_save_register_int   ("pd4990a", 0, "testwaits",    &testwaits);
+	state_save_register_int   ("pd4990a", 0, "maxwaits",     &maxwaits);
+	state_save_register_int   ("pd4990a", 0, "testbit",      &testbit);
+	state_save_register_int   ("pd4990a", 0, "outputbit",    &outputbit);
+	state_save_register_int   ("pd4990a", 0, "bitno",        &bitno);
+	state_save_register_int   ("pd4990a", 0, "reading",      &reading);
+	state_save_register_int   ("pd4990a", 0, "writting",     &writting);
+	state_save_register_int   ("pd4990a", 0, "clock_line",   &clock_line);
+	state_save_register_int   ("pd4990a", 0, "command_line", &command_line);
+}
+
 void pd4990a_addretrace()
 {
 	++testwaits;
@@ -300,9 +326,6 @@ static void pd4990a_process_command(void)
 
 void pd4990a_serial_control(unsigned char data)
 {
-	static int clock_line=0;
-	static int command_line=0;	/*??*/
-
 	/*Check for command end*/
 	if(command_line && !(data&END_BIT)) /*end of command*/
 	{
-- 
2.53.0

